python/ray/dashboard/modules/metrics/grafana_dashboard_factory.py

```diff
@@ -165,36 +173,160 @@ def _generate_grafana_dashboard(dashboard_config: DashboardConfig) -> str:
     return json.dumps(base_json, indent=4), uid
 
 
+def _generate_panel_template(
+    panel: Panel, panel_global_filters: List[str], panel_index: int, y_position: int = 0
+) -> dict:
+    """
+    Helper method to generate a panel template with common configuration.
+
+    Args:
+        panel: The panel configuration
+        panel_global_filters: List of global filters to apply
+        panel_index: The index of the panel within its row (0-based)
+        y_position: The y-coordinate of the row in the dashboard grid (0 for top-level panels)
+
+    Returns:
+        dict: The configured panel template
+    """
+    # Create base template from panel configuration
+    template = copy.deepcopy(panel.template.value)
+    template.update(
+        {
+            "title": panel.title,
+            "description": panel.description,
+            "id": panel.id,
+            "targets": _generate_targets(panel, panel_global_filters),
+        }
+    )
+
+    # Set panel position and dimensions
+    if panel.grid_pos:
+        template["gridPos"] = asdict(panel.grid_pos)
+    else:
+        # Calculate panel position in 2-column grid layout
+        # x: 0 or 12 (left or right column)
+        # y: base position + (row number * panel height)
+        template["gridPos"] = {
+            "h": PANEL_HEIGHT,
+            "w": PANEL_WIDTH,
+            "x": PANEL_WIDTH * (panel_index % PANELS_PER_ROW),
+            "y": y_position + (panel_index // PANELS_PER_ROW) * PANEL_HEIGHT,
+        }
+
+    # Configure panel visualization settings
+    template["yaxes"][0]["format"] = panel.unit
+    template["fill"] = panel.fill
+    template["stack"] = panel.stack
+    template["linewidth"] = panel.linewidth
+
+    # Handle stacking visualization
+    if panel.stack is True:
+        template["nullPointMode"] = "connected"
+
+    return template
+
+
+def _create_row_panel(row: Panel, y_position: int) -> dict:
+    """
+    Creates a Grafana row panel that spans the full dashboard width.
+    Row panels can be collapsed to hide their contained panels.
+
+    Args:
+        row: Row config with title, id, and collapse state
+        y_position: Vertical position in dashboard grid
+
+    Returns:
+        Grafana row panel configuration
+    """
+    return {
+        "collapsed": row.collapsed,
+        "gridPos": {"h": ROW_HEIGHT, "w": ROW_WIDTH, "x": 0, "y": y_position},
+        "id": row.id,
+        "title": row.title,
+        "type": "row",
+        "panels": [],
+    }
+
+
+def _calculate_panel_positions(num_panels: int, current_y: int) -> int:
+    """
+    Calculate the new y position for a set of panels.
+
+    Args:
+        num_panels: Number of panels to position
+        current_y: Current y position in the dashboard grid
+
+    Returns:
+        New y position after accounting for the panels
+    """
+    rows_needed = (num_panels + 1) // PANELS_PER_ROW
```

Comment:
why +1?

Reply:
Oops, the intention is for it to follow this:

| `num_panels` | `rows_needed` |
|--------|--------|
| 0 | 0 |
| (0, `PANELS_PER_ROW`] | 1 |
| (`PANELS_PER_ROW`, 2 * `PANELS_PER_ROW`] | 2 |

This should be the right formula....
```suggestion
    rows_needed = (num_panels + PANELS_PER_ROW - 1) // PANELS_PER_ROW
```

Reply:
nit: should we just do

`Math.ceil(num_panels // PANELS_PER_ROW)`?